### PR TITLE
Prep for client(Foo) syntax

### PR DESCRIPTION
**Pull request by mfedderly**
Converts many of the types in `@osdk/client` to work on a `ObjectOrInterfaceDefinition` instead of on a pair of `Ontology, K extends keyof Ontology["objects"]`.

Updates v2 generation to include a `type` field on each definition.

There is still remaining work to convert actions.

### CODE REVIEW

File: packages/legacy-client/src/client/objectSets/OsdkObjectSet.test.ts [hunk 1]
```diff
@@ -60,9 +60,10 @@ describe("OsdkObjectSet", () => {
 
   it("exposes descriptions", () => {
     const os = createBaseTodoObjectSet(client);
+    console.log(os);
```
Comment (mfedderly) [inline packages/legacy-client/src/client/objectSets/OsdkObjectSet.test.ts:63]:
```suggestion
```

File: packages/legacy-client/src/client/OsdkLegacyObject.test.ts [hunk 2]
```diff
@@ -22,9 +22,12 @@ import type { OsdkLegacyObjectFrom } from "./OsdkLegacyObject";
 describe("OsdkLegacyObject", () => {
   it("compiles", async () => {
     const x: Todo = {} as Todo;
-    const y: OsdkLegacyObjectFrom<typeof MockOntology, "Todo"> = x;
+    const y: OsdkLegacyObjectFrom<MockOntology, "Todo"> = x;
```
Comment (mfedderly) [inline packages/legacy-client/src/client/OsdkLegacyObject.test.ts:25]: I don't think that this is actually an example of a breaking change, but better to flag it than miss it.

File: packages/api/src/index.ts [hunk 3]
```diff
@@ -43,9 +45,11 @@ export type {
   ObjectTypeLinkDefinition,
   ObjectTypeLinkDefinitionFrom,
   ObjectTypeLinkKeysFrom,
+  ObjectTypeLinkKeysFrom2,
   ObjectTypeLinkTargetTypeFrom,
   ObjectTypePropertyDefinition,
   ObjectTypePropertyDefinitionFrom,
+  ObjectTypePropertyDefinitionFrom2,
```
[the remark this answers is not in the document]
Comment (ericanderson) [reply packages/api/src/index.ts:52]: Can't overwrite, that would be breaking changes.

File: packages/legacy-client/src/client/OsdkLegacyObject.test.ts (hunk 2 above)
Comment (ericanderson) [reply packages/legacy-client/src/client/OsdkLegacyObject.test.ts:25]: that mock is not generated, its old. this shouldnt be an issue for those using generated.

**Comment by mfedderly** [comment]
👍

**Comment by mfedderly** [comment]
LGTM but can't approve my own PR

**Pull request merged**